mono/mini/mini-llvm.c

```diff
@@ -9293,271 +9157,75 @@ mono_llvm_emit_call (MonoCompile *cfg, MonoCallInst *call)
-#endif
+	case INTRINS_MEMSET:
+		intrins = add_intrins2 (module, id, LLVMPointerType (LLVMInt8Type (), 0), LLVMInt32Type ());
 		break;
-	}
-	case INTRINS_MEMCPY: {
-#if LLVM_API_VERSION >= 900
-		/* No alignment argument */
-		LLVMTypeRef params [] = { LLVMPointerType (LLVMInt8Type (), 0), LLVMPointerType (LLVMInt8Type (), 0), LLVMInt32Type (), LLVMInt1Type () };
-
-		AddFunc (module, name, LLVMVoidType (), params, 4);
-#else
-		LLVMTypeRef params [] = { LLVMPointerType (LLVMInt8Type (), 0), LLVMPointerType (LLVMInt8Type (), 0), LLVMInt32Type (), LLVMInt32Type (), LLVMInt1Type () };
-
-		AddFunc (module, name, LLVMVoidType (), params, 5);
-#endif
+	case INTRINS_MEMCPY:
+		intrins = add_intrins3 (module, id, LLVMPointerType (LLVMInt8Type (), 0), LLVMPointerType (LLVMInt8Type (), 0), LLVMInt32Type ());
 		break;
-	}
-	case INTRINS_MEMMOVE: {
-#if LLVM_API_VERSION >= 900
-		/* No alignment argument */
-		LLVMTypeRef params [] = { LLVMPointerType (LLVMInt8Type (), 0), LLVMPointerType (LLVMInt8Type (), 0), LLVMInt64Type (), LLVMInt1Type () };
-		AddFunc (module, name, LLVMVoidType (), params, 4);
-#else
-		LLVMTypeRef params [] = { LLVMPointerType (LLVMInt8Type (), 0), LLVMPointerType (LLVMInt8Type (), 0), LLVMInt64Type (), LLVMInt32Type (), LLVMInt1Type () };
-		AddFunc (module, name, LLVMVoidType (), params, 5);
-#endif
+	case INTRINS_MEMMOVE:
+		intrins = add_intrins3 (module, id, LLVMPointerType (LLVMInt8Type (), 0), LLVMPointerType (LLVMInt8Type (), 0), LLVMInt64Type ());
 		break;
-	}
 	case INTRINS_SADD_OVF_I32:
 	case INTRINS_UADD_OVF_I32:
 	case INTRINS_SSUB_OVF_I32:
 	case INTRINS_USUB_OVF_I32:
 	case INTRINS_SMUL_OVF_I32:
-	case INTRINS_UMUL_OVF_I32: {
-		LLVMTypeRef ovf_res_i32 [] = { LLVMInt32Type (), LLVMInt1Type () };
-		LLVMTypeRef params [] = { LLVMInt32Type (), LLVMInt32Type () };
-		LLVMTypeRef ret_type = LLVMStructType (ovf_res_i32, 2, FALSE);
-
-		AddFunc (module, name, ret_type, params, 2);
+	case INTRINS_UMUL_OVF_I32:
+		intrins = add_intrins1 (module, id, LLVMInt32Type ());
 		break;
-	}
 	case INTRINS_SADD_OVF_I64:
 	case INTRINS_UADD_OVF_I64:
 	case INTRINS_SSUB_OVF_I64:
 	case INTRINS_USUB_OVF_I64:
 	case INTRINS_SMUL_OVF_I64:
-	case INTRINS_UMUL_OVF_I64: {
-		LLVMTypeRef ovf_res_i64 [] = { LLVMInt64Type (), LLVMInt1Type () };
-		LLVMTypeRef params [] = { LLVMInt64Type (), LLVMInt64Type () };
-		LLVMTypeRef ret_type = LLVMStructType (ovf_res_i64, 2, FALSE);
-
-		AddFunc (module, name, ret_type, params, 2);
+	case INTRINS_UMUL_OVF_I64:
+		intrins = add_intrins1 (module, id, LLVMInt64Type ());
 		break;
-	}
-	case INTRINS_FMA: {
-		LLVMTypeRef params [] = { LLVMDoubleType (), LLVMDoubleType (), LLVMDoubleType () };
-
-		AddFunc (module, name, LLVMDoubleType (), params, 3);
-		break;
-	}
-	case INTRINS_FMAF: {
-		LLVMTypeRef params [] = { LLVMFloatType (), LLVMFloatType (), LLVMFloatType () };
-
-		AddFunc (module, name, LLVMFloatType (), params, 3);
-		break;
-	}
+	case INTRINS_FMA:
```

Comment:
Wouldn't `INTRINS_FMA` need `add_intrins3 (module, id, LLVMDoubleType (), LLVMDoubleType (), LLVMDoubleType ())`?

Reply:
It only needs the extra parameters, not the whole signature, i.e. in llvm its named
llvm.fma.f64.

Reply:
Ah, got it.

(https://github.com/dotnet/llvm-project/blob/release/6.x/llvm/lib/IR/Function.cpp#L597)